Docs for External Sessions Data store (#1225)
SeleniumHQ/seleniumhq.github.io: website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md

---
title: "External datastore for sessions"

> linkTitle: "External datastore for sessions"

Reviewer: ```suggestion
title: "External datastore"
linkTitle: "External datastore"
```

Reviewer: Let's leave it "External datastore" only, as the goal in the future is to have this also for the Distributor. So both things could go here. Also the file name could change from `external_datastore_sessions.en.md` to `external_datastore.en.md`

Author: Fixed

## Table of Contents
* [Why an external datastore](#why-an-external-datastore)
* [The solution](#the-solution)
* [Setup](#setup)
* [Database backed Session Map](#database-backed-session-map) 
    * [Steps](#steps)
* [Redis backed Session Map](#redis-backed-session-map) 
    * [Steps](#steps)

## Why an external datastore ?

When you are running a distributed grid that comprises of:

* Event Bus
* New Session Queue
* Session Map
* Distributor
* Router
* Node

you may want to scale these individual components horizontally (i.e., run more than one instances of the same component. For e.g., two `Session Map`) to cater to your grid's increased load.

What would happen if we run two instances of a `Session Map`? 

By default, a `Session Map` component uses an in-memory data store to store sessions information (which session is running on which node).

So if there are two `Session Map` instances, then we end up with 2 in-memory data store instances which are not shared.

This can cause a lot of issues such as:

* Sessions data not being shared by both the instances
* Session data query goes to the wrong instance of `Session Map`
* Sessions data is lost if one of the instances goes down

> In short we now have a consistency problem that is so closely associated with the notion of microservices.

Reviewer: With the current Grid 4, this is not accurate because it is not possible to run two `SessionMap` instances. Well, it is technically possible, but the `Router` can only talk to only one `SessionMap` instance.

The reasoning behind backing the data from the `SessionMap` is to allow the user to take down the `SessionMap` instance and bring a new one without loosing the current data.

As you can see, the topic of a high availability Grid has lots of room for improvement.

Author: Wouldn't it work fine if we put up a reverse proxy/load balancer (maybe something like a haproxy or nginx) which just does routing on a round robin basis? In that case, the router would just connect to the load balancer that fronts the session map and so it should be possible no ?

Reviewer: Yes, but that is out of the scope of this doc for now.

That is what I tried to say with
>As you can see, the topic of a high availability Grid has lots of room for improvement.

We need to figure out ways for users to have a high availability Grid. What you propose is an alternative.

Author: > We need to figure out ways for users to have a high availability Grid. 

I feel that if Grid helps users to leverage High Availability Tools (which are meant for that purpose) then we wouldn't need to build them (I hope that we aren't looking at that) and so the ask of high availability would have been resolved right ?

Reviewer: Yes, we should reuse existing tools as much as possible. But as the current state, we can only document what we know that works. 

A future step would be to figure out what tools are needed and how to combine them with Grid to make it highly available. 

We can continue this conversation in the #selenium-tlc channel on Slack 🙏

Author: Fixed

## The solution

> The simplest solution would be to ensure that all the sessions map instances use a common shared data store. And thus we have a need to externalise this data store which holds running sessions information.

Reviewer: This needs to be reworked based on the previous comment.

Author: Fixed

> Thanksfully Grid4 is already geared up to support this use case. 

Reviewer: ```suggestion
Thankfully Grid already supports this use case. 
```

Author: Fixed

> You can use a data store that is: 
> 
> * Backed by your favourite database (or) 
> * Backed by a Redis Cache system
> 
> so that all the running sessions map instances use a single shared data store.

Reviewer: ```suggestion
You can use a data store that is backed by your favorite database or by a Redis cache system.
```

Author: Fixed

## Database backed Session Map

For the sake of this illustration, we are going to work with PostGreSQL database.

We will spin off a PostGreSQL database as a docker container using a docker compose file.

### Steps

You can skip this step if you already have a PostGreSQL database instance available at your disposal.

* Create a sql file named `init.sql` with the below contents:

```sql
CREATE TABLE IF NOT EXISTS sessions_map(
    session_ids varchar(256),
    session_caps text,
    session_uri varchar(256),
    session_stereotype text,
    session_start varchar(256)
 );
```

* In the same directory as the `init.sql`, create a file named `docker-compose.yml` with its contents as below:

```yaml
version: '3.8'
services:
  db:
    image: postgres:9.6-bullseye
    restart: always
    environment:
      - POSTGRES_USER=seluser
      - POSTGRES_PASSWORD=seluser
      - POSTGRES_DB=selenium_sessions
    ports:
      - "5432:5432"
    volumes:
    - ./init.sql:/docker-entrypoint-initdb.d/init.sql
```

We can now start our database container by running:

```bash
docker-compose up -d
```

> _So our database name is `selenium_sessions` with its username and password set to `seluser`_

Reviewer: ```suggestion
_Our database name is `selenium_sessions` with its username and password set to `seluser`_
```

Author: Fixed

If you are working with an already running PostGreSQL DB instance, then you just need to create a database named `selenium_sessions` and the table `sessions_map` using the above mentioned SQL statement.

* Create a Selenium configuration file named `sessions.toml` with the below contents:

```toml
[sessions]
implementation = "org.openqa.selenium.grid.sessionmap.jdbc.JdbcBackedSessionMap"
jdbc-url = "jdbc:postgresql://localhost:5432/selenium_sessions"
jdbc-user = "seluser"
jdbc-password = "seluser"
```

_Note:_ If you plan to use an existing PostGreSQL DB instance, then replace `localhost:5432` with the actual host and port number of your instance.

> * Below is a simple shell script (let's call it `distributed.sh`) that we will use to bring up our distributed grid.

Reviewer: ```suggestion
* Below is a simple shell script (let's call it `distributed.sh`) that we will use to bring up our distributed Grid.
```

Author: Fixed

```bash

> SE_VERSION=4.6.0

Reviewer: ```suggestion
SE_VERSION=<current_selenium_version>
```

Reviewer: I prefer to leave this placeholder because it makes the docs easier to maintain.

Author: Fixed

JAR_NAME=selenium-server-${SE_VERSION}.jar
PUBLISH="--publish-events tcp://localhost:4442"
SUBSCRIBE="--subscribe-events tcp://localhost:4443"
SESSIONS="--sessions http://localhost:5556"
SESSIONS_QUEUE="--sessionqueue http://localhost:5559"

> # Start event bus

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

echo 'Starting Event Bus'
java -jar $JAR_NAME event-bus $PUBLISH $SUBSCRIBE --port 5557 &

> #Start New Session Queue

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

echo 'Starting New Session Queue'
java -jar $JAR_NAME sessionqueue --port 5559 &
#Start session map

> echo 'Starting sessions manager'

Reviewer: ```suggestion
echo 'Starting Session Map'
```

Author: Fixed

java -jar $JAR_NAME \
--ext $(coursier fetch -p org.seleniumhq.selenium:selenium-session-map-jdbc:${SE_VERSION} org.postgresql:postgresql:42.3.1) \
sessions $PUBLISH $SUBSCRIBE --port 5556 --config sessions.toml &

> #Start distributor

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

> echo 'Starting distributor'

Reviewer: ```suggestion
echo 'Starting Distributor'
```

Author: Fixed

java -jar $JAR_NAME  distributor $PUBLISH $SUBSCRIBE $SESSIONS $SESSIONS_QUEUE --port 5553 --bind-bus false &

> #Start router

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

> echo 'Starting router'

Reviewer: ```suggestion
echo 'Starting Router'
```

Author: Fixed

java -jar $JAR_NAME router $SESSIONS --distributor http://localhost:5553 $SESSIONS_QUEUE --port 4444 &

> #Start node

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

> echo 'Starting node'

Reviewer: ```suggestion
echo 'Starting Node'
```

Author: Fixed

java -jar $JAR_NAME node $PUBLISH $SUBSCRIBE &
```

* At this point the current directory should contain the following files:
    * `docker-compose.yml` 
    * `init.sql`
    * `sessions.toml`
    * `distributed.sh`

> * You can now spawn the grid by running `distributed.sh` shell script and quickly run a test. You will notice that the Grid now stores (inserts and then deletes) session information into the PostGreSQL database.

Reviewer: ```suggestion
* You can now spawn the Grid by running `distributed.sh` shell script and quickly run a test. You will notice that the Grid now stores session information into the PostGreSQL database.
```

Author: Fixed

> In the line which spawns a `Session Map` on a machine:

Reviewer: ```suggestion
In the line which spawns a `SessionMap` on a machine:
```

Author: Fixed

```bash

> export SE_VERSION=4.6.0

Reviewer: ```suggestion
export SE_VERSION=<current_selenium_version>
```

Author: Fixed

--port 5556 --config sessions.toml 
```

* The variable names from the above script have been replaced with their actual values for clarity.
* Remember to substitute `localhost` with the actual hostname of the machine where your `Event-Bus` is running.
* The arguments being passed to `coursier` are basically the GAV (Group Artifact Version) Maven co-ordinates of: 
    * [selenium-session-map-jdbc](https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-session-map-jdbc) which is needed to help us store sessions information in database
    * [postgresql](https://mvnrepository.com/artifact/org.postgresql/postgresql) which is needed to help us talk PostGreSQL database.
* `sessions.toml` is the configuration file that we created earlier.

## Redis backed Session Map

We will spin off a Redis Cache docker container using a docker compose file.

### Steps

You can skip this step if you already have a Redis Cache instance available at your disposal.

* Create a file named `docker-compose.yml` with its contents as below:

```yaml
version: '3.8'
services:
  redis:
    image: redis:bullseye
    restart: always
    ports:
      - "6379:6379"
```

> We can now start our redis container by running:

Reviewer: ```suggestion
We can now start our Redis container by running:
```

Author: Fixed

```bash
docker-compose up -d
```

> * Create a Selenium configuration file named `sessions.toml` with the below contents:

Reviewer: ```suggestion
* Create a Selenium Grid configuration file named `sessions.toml` with the below contents:
```

Author: Fixed

```toml
[sessions]
scheme = "redis"
implementation = "org.openqa.selenium.grid.sessionmap.redis.RedisBackedSessionMap"
hostname = "localhost"
port = 6379
```

_Note:_ If you plan to use an existing Redis Cache instance, then replace `localhost` and `6379` with the actual host and port number of your instance.

* Below is a simple shell script (let's call it `distributed.sh`) that we will use to bring up our distributed grid.

```bash

> SE_VERSION=4.6.0

Reviewer: ```suggestion
SE_VERSION=<current_selenium_session>
```

Reviewer: I prefer to have a placeholder to simplify docs maintenance.

Author: Fixed

JAR_NAME=selenium-server-${SE_VERSION}.jar
PUBLISH="--publish-events tcp://localhost:4442"
SUBSCRIBE="--subscribe-events tcp://localhost:4443"
SESSIONS="--sessions http://localhost:5556"
SESSIONS_QUEUE="--sessionqueue http://localhost:5559"

> # Start event bus

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

echo 'Starting Event Bus'
java -jar $JAR_NAME event-bus $PUBLISH $SUBSCRIBE --port 5557 &

> #Start New Session Queue

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

echo 'Starting New Session Queue'
java -jar $JAR_NAME sessionqueue --port 5559 &

> #Start session map

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

> echo 'Starting sessions manager'

Reviewer: ```suggestion
echo 'Starting Session Map'
```

Author: Fixed

java -jar $JAR_NAME \
--ext $(coursier fetch -p org.seleniumhq.selenium:selenium-session-map-redis:${SE_VERSION}) \
sessions $PUBLISH $SUBSCRIBE --port 5556 --config sessions.toml &

> #Start distributor

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

> echo 'Starting distributor'

Reviewer: ```suggestion
echo 'Starting Distributor'
```

Author: Fixed

java -jar $JAR_NAME  distributor $PUBLISH $SUBSCRIBE $SESSIONS $SESSIONS_QUEUE --port 5553 --bind-bus false &

> #Start router

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

> echo 'Starting router'

Reviewer: ```suggestion
echo 'Starting Router'
```

Author: Fixed

java -jar $JAR_NAME router $SESSIONS --distributor http://localhost:5553 $SESSIONS_QUEUE --port 4444 &

> #Start node

Reviewer: ```suggestion
```

Reviewer: No need to write this as the `echo` from the next line already says the same.

Author: Fixed

> echo 'Starting node'

Reviewer: ```suggestion
echo 'Starting Node'
```

Author: Fixed

java -jar $JAR_NAME node $PUBLISH $SUBSCRIBE &
```

* At this point the current directory should contain the following files:
    * `docker-compose.yml`
    * `sessions.toml`
    * `distributed.sh`

> * You can now spawn the grid by running `distributed.sh` shell script and quickly run a test. You will notice that the Grid now stores (inserts and then deletes) session information into the Redis instance. You can perhaps make use of a Redis GUI such as [TablePlus](https://tableplus.com/) to see them (Make sure that you have setup a debug point in your test, because the values will get deleted as soon as the test runs to completion)

Reviewer: ```suggestion
* You can now spawn the Grid by running `distributed.sh` shell script and quickly run a test. You will notice that the Grid now stores session information into the Redis instance. You can use a Redis GUI such as [TablePlus](https://tableplus.com/) to see them (make sure that you have setup a debug point in your test, because the values will get deleted as soon as the test runs to completion).
```

Author: Fixed

In the line which spawns a `Session Map` on a machine:

```bash

> export SE_VERSION=4.6.0

Reviewer: ```suggestion
export SE_VERSION=<current_selenium_version>
```

Author: Fixed
